pkg/controller/node/node_controller.go

```diff
@@ -1042,3 +1064,108 @@ func getErrorString(err error) string {
+			ctrlcommon.MachineConfigControllerPausedPoolKubeletCA.WithLabelValues(pool.Name).Set(float64(newestSignerCertificate.NotAfter.UTC().Unix()))
+		}
+		break
+	}
+}
+
+// resetImportantFileAlerts turns off any "paused file" alerts that were firing for the pool
+func (ctrl *Controller) resetImportantFileAlerts(pool *mcfgv1.MachineConfigPool) {
+	// Set the alert for this pool back to zero
+	ctrlcommon.MachineConfigControllerPausedPoolKubeletCA.WithLabelValues(pool.Name).Set(0)
+}
+
+// parseConvertMachineConfigFilesForPool retrieves the current and pending configurations for a pool, parses and converts them, and
+// returns them as ignition v3 Config objects
+func (ctrl *Controller) parseConvertMachineConfigFilesForPool(pool *mcfgv1.MachineConfigPool) (current, pending *ign3types.Config, err error) {
+	// The config we're in right now
+	currentName := pool.Status.Configuration.Name
+	// The config we would be going to
+	pendingName := pool.Spec.Configuration.Name
+
+	// Get the machine config objects
+	currentConfig, err := ctrl.mcLister.Get(currentName)
+	if apierrors.IsNotFound(err) {
+		glog.V(2).Infof("MachineConfig %v has been deleted", currentName)
+		return nil, nil, err
+	}
+
+	pendingConfig, err := ctrl.mcLister.Get(pendingName)
+	if apierrors.IsNotFound(err) {
+		glog.V(2).Infof("MachineConfigPool %v has been deleted", pendingName)
+		return nil, nil, err
+	}
+
+	// Make sure we can coax the objects into ignitionv3
+	currentIgnConfig, err := ctrlcommon.ParseAndConvertConfig(currentConfig.Spec.Config.Raw)
+	if err != nil {
+		return nil, nil, err
+	}
+	pendingIgnConfig, err := ctrlcommon.ParseAndConvertConfig(pendingConfig.Spec.Config.Raw)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &currentIgnConfig, &pendingIgnConfig, nil
+}
+
+// getNewestAPIToKubeletSignerCertificate returns the newest kube-api-to-kubelet-signer certificate present in the
+// kubelet-ca.crt bundle
+func (ctrl *Controller) getNewestAPIToKubeletSignerCertificate(statusIgnConfig *ign3types.Config) (*x509.Certificate, error) {
+	// Retrieve the file data from ignition
+	kubeletBundle, err := ctrlcommon.GetIgnitonFileDataByPath(statusIgnConfig, kubeletCAFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse that bundle into its component certificates
+	containedCertificates, err := ctrlcommon.GetCertificatesFromPEMBundle(kubeletBundle)
+	if err != nil {
+		return nil, err
+	}
+
+	// We have other problems if this is empty, but it's possible
+	if len(containedCertificates) == 0 {
+		return nil, fmt.Errorf("No certificates found in bundle")
+	}
+
+	// The *original* signer has a different name, the rotated ones have longer names
+	// The suffix changes with the timstamp on rotation, which is why I'm using prefix here not exact match
+	newestCertificate := ctrlcommon.GetLongestValidCertificate(containedCertificates, kubeAPIToKubeletSignerNamePrefixes)
```

Comment:
One question: Does looking for longest valid cert mean that we will not be sending critical alert for cert rotation that occurs at 365 days to delete old expired cert?

Reply:
Correct, it would not go critical under that circumstance because the certificates would have too many days remaining. 

 - Do we consider the rotation for removal of that certificate critical? ( And would someone be able to quantify the bad thing that happens if that can't take place?) 
 
I can figure out if we're in that condition by doing something like "If the pending machine configs removed certificates from this bundle, and we still have plenty of days remaining, then this is the the 'cleanup' rotation".

Reply:
nice, this is expected behavior. Was just making sure that we are not going critical during cert removal change.